### DISCUSSION TITLE
Tanstack Router + Relay

### DISCUSSION THREAD

**Discussion by ValenCassa** [TanStack/router#807] [Q&A]
Hi all! I'm using Tanstack router + Relay for my app but now that loaders are gone, is it possible to still pass in the `initialQueryRef` that you get from `loadQuery` to the component?

**Comment by tannerlinsley** [comment]
There is no storage or linkage between the load function and the component now. Just the suspense. You’ll need to store any data or refs outside the component for now. I could possibly wire up a context…

**Reply by ValenCassa** [reply]
Thanks for answering Tanner! I'd really love to continue using the library but not being able to pass in my refs to the components just overcomplicates things in my case, but I recognize this one is more like an edge case

**Reply by tannerlinsley** [reply]
Can you show me a pseudo code example of what you’d like to do? I bet I can make it work for you using relay is one of my objectives as well.I’m

**Reply by ValenCassa** [reply]
I'd like to be able to do something similar to what the loader function did or what it does in `react-router`. Could be something like:
```jsx
const Query = graphql`...`
const Component = () => {
  const {initialQueryRef} = route.useRouteContext()
  const data = useQueryLoader(Query, initialQueryRef)
  
  return {...}
}

const route = new Route({
  // ops: context, params, etc
  beforeRender: (ops) => {
    return {
      initialQueryRef: loadQuery(relayEnvironment, Query)
    }
  }
  ...
})
```

**Reply by tannerlinsley** [reply]
Gimme a day or two

**Reply by MarkArchitect** [reply]
Hi, due to the fact that Loaders have been removed, I also have a problem with the current version of the router and relay.
Below the code I used. Is there any solution to this problem?

Definition of route with loaders:




```typescript
export const FilmDetailRoute = new Route({
    getParentRoute: () => rootRoute,
    path: 'filmdetails/$filmId',
    loader:async({params:{filmId}})=>{
            const filmDetailQueryRef= await loadQuery<filmDetailQueryType>(
            RelayEnvironment,
            filmDetailQuery,
            {
              id: filmId,
            },
            { fetchPolicy: 'store-or-network' },
          )
          return filmDetailQueryRef;
    },
  }).update({
    component: lazyRouteComponent(() => import('../components/film-detail'), 'FilmDetail'),
  })
```

definition of component which consumes preloaded data from router loader


```typescript
export const FilmDetail: (typeof FilmDetailRoute)["options"]["component"] = ({
  useLoader,
  useParams,
}) => {
  const params = useParams();

  const preData = useLoader();

  const { film } = usePreloadedQuery<filmDetailQuery>(
    graphql`
      query filmDetailQuery($id: ID!) {
        film(id: $id) {
          id
          title
          director
          releaseDate
          created
          ...characterListFragment @arguments(showMore: true)
          ...characterList2Fragment @arguments(showMore: false)
          ...characterListFragmentTest
          ...characterList3Fragment
        }
      }
    `,
    preData
  );
```

film has all the data

I will be grateful for any help

**Reply by tannerlinsley** [reply]
I’m working on a solution.

**Reply by ValenCassa** [reply]
@tannerlinsley Hey Tanner! Are there any updates on this?

**Reply by tannerlinsley** [reply]
Yes, loaders are back!

**Comment by sairion** [comment]
it would be nice if I can build Tanstack Router version of https://github.com/loop-payments/react-router-relay (long story short, you can use EntryPoint pattern with it)
